### PR TITLE
allow convert error to be handled

### LINKED ISSUE #905: fallible ConvertBalance
make `convert_balance` and `convert_balance_back` return `Result` to be able to handle overflow

### PR DESCRIPTION
**Pull request by xlc**
closes #905

### CODE REVIEW

**Review by ermalkaleci — APPROVED** [review]

File: tokens/src/impls.rs [hunk 1]
```diff
@@ -146,43 +162,48 @@ where
...
-			C::convert_balance_back(amount, GetCurrencyId::get()),
-		);
+
+		let amount = C::convert_balance_back(amount, GetCurrencyId::get());
+		let amount = match amount {
+			Ok(amount) => amount,
+			Err(ArithmeticError::Overflow) => return Overflow,
+			Err(ArithmeticError::Underflow) => return Underflow,
+			Err(ArithmeticError::DivisionByZero) => return Overflow,
+		};
```
Comment (zjb0807) [inline tokens/src/impls.rs:201]: tag1

File: tokens/src/impls.rs [hunk 2]
```diff
@@ -146,43 +162,48 @@ where
...
-			GetCurrencyId::get(),
-			who,
-			C::convert_balance_back(amount, GetCurrencyId::get()),
-			mint,
-		)
+		let amount = C::convert_balance_back(amount, GetCurrencyId::get());
+		let amount = match amount {
+			Ok(amount) => amount,
+			Err(_) => return DepositConsequence::Overflow,
+		};
```
Comment (zjb0807) [inline tokens/src/impls.rs:188]: Why is this different from tag1 below?

**Review by zjb0807 — APPROVED** [review]
LGTM

File: tokens/src/impls.rs (hunk 2 above)
Comment (xlc) [reply tokens/src/impls.rs:188]: there is `WithdrawConsequence::Underflow` but no `DepositConsequence::Underflow`
https://github.com/paritytech/substrate/blob/c0793b58c509f352b6006e402350ac9ad8017823/frame/support/src/traits/tokens/misc.rs#L93

**Pull request merged**